File: src/types/printer.ts
import { type DirectoryEntry } from "@/types/directory";

export interface PrinterInfo {
  dn: string;
  name: string;
  location: string;
  serverName: string;
  sharePath: string;
  driverName: string;
  description: string;
}

export function mapEntryToPrinter(entry: DirectoryEntry): PrinterInfo {
  const attr = (name: string): string => entry.attributes[name]?.[0] ?? "";
  return {
    dn: entry.distinguishedName,
    name: attr("printerName") || entry.displayName || entry.samAccountName || "",
    location: attr("location"),
    serverName: attr("serverName"),
    sharePath: attr("uNCName"),
    driverName: attr("driverName"),
    description: attr("description"),
  };
}

File: src-tauri/src/models/printer.rs
use serde::{Deserialize, Serialize};

/// Represents an AD printer (printQueue) object.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrinterInfo {
    /// The distinguished name of the printer.
    pub dn: String,
    /// Printer name (printerName / cn).
    pub name: String,
    /// Physical location (location).
    pub location: String,
    /// Print server name (serverName).
    pub server_name: String,
    /// UNC share path (uNCName).
    pub share_path: String,
    /// Driver name (driverName).
    pub driver_name: String,
    /// Description.
    pub description: String,
}
